presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -189,6 +218,78 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
                     true));
         }
 
+        private Optional<DecorrelationResult> rewriteLimitWithRowCountGreaterThanOne(DecorrelationResult childDecorrelationResult, long limit, PlanNodeId nodeId)
+        {
+            PlanNode decorrelatedChildNode = childDecorrelationResult.node;
+            Set<Symbol> constantSymbols = childDecorrelationResult.getConstantSymbols();
+
+            if (constantSymbols.isEmpty() || !constantSymbols.containsAll(childDecorrelationResult.symbolsToPropagate)) {
+                return Optional.empty();
+            }
+
+            // rewrite Limit to RowNumberNode partitioned by constant symbols
+            RowNumberNode rowNumberNode = new RowNumberNode(
+                    nodeId,
+                    decorrelatedChildNode,
+                    ImmutableList.copyOf(childDecorrelationResult.symbolsToPropagate),
+                    symbolAllocator.newSymbol("row_number", BigintType.BIGINT),
+                    Optional.of(toIntExact(limit)),
+                    Optional.empty());
+
+            return Optional.of(new DecorrelationResult(
+                    rowNumberNode,
+                    childDecorrelationResult.symbolsToPropagate,
+                    childDecorrelationResult.correlatedPredicates,
+                    childDecorrelationResult.correlatedSymbolsMapping,
+                    false));
+        }
+
+        @Override
+        public Optional<DecorrelationResult> visitTopN(TopNNode node, Void context)
+        {
+            if (node.getCount() == 0) {
+                return Optional.empty();
+            }
+
+            Optional<DecorrelationResult> childDecorrelationResultOptional = lookup.resolve(node.getSource()).accept(this, null);
+            if (!childDecorrelationResultOptional.isPresent()) {
+                return Optional.empty();
+            }
+
+            DecorrelationResult childDecorrelationResult = childDecorrelationResultOptional.get();
+            if (childDecorrelationResult.atMostSingleRow) {
+                return childDecorrelationResultOptional;
+            }
+
+            PlanNode decorrelatedChildNode = childDecorrelationResult.node;
+            Set<Symbol> constantSymbols = childDecorrelationResult.getConstantSymbols();
+
+            if (constantSymbols.isEmpty() || !constantSymbols.containsAll(childDecorrelationResult.symbolsToPropagate)) {
+                return Optional.empty();
```

Comment:
nit: in such cases we could rewrite to `TopNRowNumberNode` above decorrelated join

Reply:
Yes, but it would also require assigning unique ids to input rows and adding them to `TopNRowNumberNode`'s partitioning symbols to avoid equal input rows sharing subquery results.